### PR TITLE
Fixup some bugs found after merging atop fusion

### PR DESCRIPTION
**Pull request by mrocklin**
These are causing failures in the dask-ml test suite.

cc @TomAugspurger

### CODE REVIEW

**Comment by mrocklin** [comment]
There's more to do here, but we're getting closer

**Comment by mrocklin** [comment]
@TomAugspurger this may resolve dask-ml things.  Things seem to pass locally for me, but I'd appreciate it if you could also check on your end.

**Comment by TomAugspurger** [comment]
Hmm I'm getting sporadic failures on `pytest tests/test_svd.py::test_integers` in dask-ml.

**Comment by TomAugspurger** [comment]
```
tests/test_svd.py::test_integers FAILED                                                                                                              [100%]

========================================================================= FAILURES =========================================================================
______________________________________________________________________ test_integers _______________________________________________________________________

    def test_integers():
        Xint = dXdense.astype(np.int64)
        tsvd = dd.TruncatedSVD(n_components=6)
>       Xtrans = tsvd.fit_transform(Xint)

tests/test_svd.py:87:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dask_ml/decomposition/truncated_svd.py:181: in fit_transform
    components, ev, evr, sv = compute(v, explained_var, explained_variance_ratio, s)
../dask/dask/base.py:393: in compute
    results = schedule(dsk, keys, **kwargs)
../dask/dask/threaded.py:75: in get
    pack_exception=pack_exception, **kwargs)
../dask/dask/local.py:501: in get_async
    raise_exception(exc, tb)
../dask/dask/compatibility.py:112: in reraise
    raise exc
../dask/dask/local.py:272: in execute_task
    result = _execute_task(task, data)
../dask/dask/local.py:253: in _execute_task
    return func(*args2)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = ('astype-78945f59659a0976e5e7f15deeb31fe5', 0, 0)

    def _wrapped_qr(a):
        """
        A wrapper for np.linalg.qr that handles arrays with 0 rows

        Notes: Created for tsqr so as to manage cases with uncertain
        array dimensions. In particular, the case where arrays have
        (uncertain) chunks with 0 rows.
        """
        # workaround may be removed when numpy stops rejecting edge cases
>       if a.shape[0] == 0:
E       AttributeError: 'tuple' object has no attribute 'shape'

../dask/dask/array/linalg.py:45: AttributeError
```

**Comment by mrocklin** [comment]
OK, tests in dask, dask-ml, and xarray pass

**Comment by mrocklin** [comment]
This is blocking tests in dask-ml.  Pushing through quickly.  Merging on a green light from CI.

File: dask/array/linalg.py [hunk 1]
```diff
@@ -183,10 +185,12 @@ def tsqr(data, compute_svd=False, _max_vchunk_size=None):
                                            [(name_r_st1, idx, 0)
                                             for idx, _ in sub_block_info])))
                              for i, sub_block_info in enumerate(all_blocks))
-        dsk.update_with_key(dsk_r_stacked, key=name_r_stacked)
+        dsk.update_with_key(dsk_r_stacked, key=name_r_stacked,
+                            dependencies={name_r_st1})
 
         # retrieve R_stacked for recursion with tsqr
         vchunks_rstacked = tuple([sum(map(lambda x: x[1], sub_block_info)) for sub_block_info in all_blocks])
+        dsk.dependencies[name_r_stacked] = set(data.name)
```
Comment (TomAugspurger) [inline dask/array/linalg.py]: Is `data.name` a string? Should this instead be `set([data.name])`?

File: dask/array/linalg.py [hunk 2]
```diff
@@ -327,6 +343,8 @@ def tsqr(data, compute_svd=False, _max_vchunk_size=None):
             r_shape = (n, n) if data.shape[0] >= data.shape[1] else data.shape
             r_chunks = r_shape
 
+        dsk.dependencies[name_q_st3] = set(data.name)
```
Comment (TomAugspurger) [inline dask/array/linalg.py]: A couple more `set(data.name)` (which may not be a problem) 
https://github.com/dask/dask/pull/4081/files#r224559780

File: dask/array/linalg.py (hunk 1 above)
Comment (TomAugspurger) [reply dask/array/linalg.py]: Or rather `{data.name}`

File: dask/sharedict.py [hunk 3]
```diff
@@ -53,7 +53,10 @@ def __init__(self, dicts=None, dependencies=None):
         self.dicts = dicts or dict()
         self.dependencies = dependencies or dict()
 
-    def update_with_key(self, arg, key=None):
+        assert set(self.dependencies) == set(self.dicts)
+        # assert all(dep in self.dicts for v in self.dependencies.values() for dep in v)
```
Comment (TomAugspurger) [inline dask/sharedict.py]: Any value in keeping this comment around?

**Comment by TomAugspurger** [comment]
One potential bug. +1 otherwise

File: dask/array/linalg.py (hunk 1 above)
Comment (mrocklin) [reply dask/array/linalg.py]: Thank you for the catch.  This was, indeed, a bug.  

I've added an assert statement for now to catch these in the future.  I suspect that there will be some churn in this area soonish.

File: dask/sharedict.py (hunk 3 above)
Comment (mrocklin) [reply dask/sharedict.py]: removed

**Comment by TomAugspurger** [comment]
Can't reproduce locally, but the one I spot-checked looks like floating point error:

```
nan, 0.10974073322118284, nan,
nan, 0.10974073322127739, nan,
```

**Pull request merged**